logger/logger.go

```diff
@@ -51,44 +55,59 @@ type Message struct {
-		os.Exit(3)
-	}
+// InvalidOptionMessage represents an issue with the supplied configuration.
+type InvalidOptionMessage struct {
+	Option  string
+	Message string
+}
+
+func (e InvalidOptionMessage) String() string {
+	return fmt.Sprintf("Options.%s: %s", e.Option, e.Message)
 }
 
-func checkOptions(options *Options) {
+func (options *Options) validate() error {
+	var problems []string
+	reMacAddress := regexp.MustCompile(`^([0-9a-fA-F][0-9a-fA-F]:){5}([0-9a-fA-F][0-9a-fA-F])`)
+	reHostname := regexp.MustCompile(`(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9-]*[A-Za-z0-9])`)
+	reIPAddress := regexp.MustCompile(`(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}`)
+
+	validateOptionLength("App", options.App, &problems)
+	validateOptionLength("Env", options.Env, &problems)
+	validateOptionLength("Hostname", options.Hostname, &problems)
+	validateOptionLength("Level", options.Level, &problems)
+
+	if options.MacAddress != "" && (!reMacAddress.MatchString(options.MacAddress)) {
+		problems = append(problems, InvalidOptionMessage{"MacAddress", "invalid format"}.String())
+	}
+
+	if options.Hostname != "" && !reHostname.MatchString(options.Hostname) {
+		problems = append(problems, InvalidOptionMessage{"Hostname", "invalid format"}.String())
+	}
+
+	if options.IPAddress != "" && !reIPAddress.MatchString(options.IPAddress) {
+		problems = append(problems, InvalidOptionMessage{"IPAddress", "invalid format"}.String())
+	}
+
 	if options.FlushInterval == 0 {
 		options.FlushInterval = 5 * time.Second
 	}
 	if options.MaxBufferLen == 0 {
 		options.MaxBufferLen = 5
 	}
 	if options.IngestURL == "" {
-		options.IngestURL = "https://logs.logdna.com/logs/ingest"
+		options.IngestURL = defaultIngestURL
 	}
 
-	checkParameterLength("App", options.App)
-	checkParameterLength("Env", options.Env)
-	checkParameterLength("Hostname", options.Hostname)
-	checkParameterLength("Level", options.Level)
+	if len(problems) > 0 {
+		return errors.New(strings.Join(problems, ", "))
+	}
 
-	reMacAddress := regexp.MustCompile(`^([0-9a-fA-F][0-9a-fA-F]:){5}([0-9a-fA-F][0-9a-fA-F])`)
-	reHostname := regexp.MustCompile(`(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])\.)*([A-Za-z0-9]|[A-Za-z0-9][A-Za-z0-9-]*[A-Za-z0-9])`)
-	reIPAddress := regexp.MustCompile(`(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)(\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)){3}`)
+	return nil
+}
 
-	if options.MacAddress != "" && (!reMacAddress.MatchString(options.MacAddress)) {
-		fmt.Println("Invalid MAC Address format.")
-		os.Exit(3)
-	}
-	if options.Hostname != "" && !reHostname.MatchString(options.Hostname) {
-		fmt.Println("Invalid hostname.")
-		os.Exit(3)
-	}
-	if options.IPAddress != "" && !reIPAddress.MatchString(options.IPAddress) {
-		fmt.Println("Invalid IP Address format.")
-		os.Exit(3)
+func validateOptionLength(option string, value string, problems *[]string) {
+	if len(value) > 32 {
```

Comment:
The 32 limit is going to be inconsistent with the new node logger changes which set the input limit at 80: https://github.com/logdna/logger-node/pull/1

If we want to enforce limits on the client side I think we should extend from the limits as defined in our documented Ingest API: https://docs.logdna.com/docs/ingestion#service-limits

This might warrant a discussion about what our limits should be and whether that is accurately reflected in our documentation, ingester api, and clients.

Reply:
i'll add this as a task in LOG-6969 where validation will be reworked. agreed though, and we'll want to decide for each of those documented limits where they should be enforced.